File: wallet-android/src/main/java/com/iohk/jormungandrwallet/Wallet.kt
package com.iohk.jormungandrwallet

class Wallet {
    private var walletPtr: Long = 0


    private external fun recover(mnemonics: String) : Long
    private external fun delete(wallet: Long)
    private external fun totalValue(wallet: Long) : Int
    private external fun initialFunds(wallet: Long, block0: ByteArray) : Long
    companion object {
        init {
            System.loadLibrary("wallet_jni")
        }
    }

    constructor(mnemonics: String) {
        this.walletPtr = recover(mnemonics)
        assert(this.walletPtr != 0.toLong())
    }

    fun initialFunds(block0: ByteArray) : Settings {
        val settings_ptr : Long = initialFunds(this.walletPtr, block0)
        return Settings(settings_ptr)
    }

    fun value() : Int {
        return totalValue(this.walletPtr)
    }
}

File: bindings/wallet-swift/Sources/JormungandrWallet/Wallet.swift
import Foundation
import JormungandrWalletC

class Wallet {
    private var pointer: WalletPtr

    init(withMnemonics mnemonics: String) throws {
        self.pointer = try WalletC.Wallet.recover(mnemonics: mnemonics)
    }

    func totalValue() throws -> UInt64 {
        return try WalletC.Wallet.totalValue(wallet: self.pointer)
    }

    func settings(block0: Data) throws -> Settings {
        return try Settings(
            withRawPointer: WalletC.Wallet.settings(wallet: self.pointer, block0: block0)
        )
    }

    func id() throws -> Data {
        return try WalletC.Wallet.id(wallet: self.pointer)
    }

    func setState(value: UInt64, counter: UInt32) throws {
        try WalletC.Wallet.setState(wallet: self.pointer, value: value, counter: counter)
    }

    func vote(settings: Settings, proposal: Proposal, choice: UInt8) throws -> Data {
        return try WalletC.Wallet.castVote(
            wallet: self.pointer,
            settings: settings.pointer,
            proposal: proposal.pointer,
            choice: choice
        )
    }

    func convert(settings: Settings) throws -> Conversion {
        return try Conversion(
            withRawPointer: WalletC.Wallet.convert(wallet: self.pointer, settings: settings.pointer)
        )
    }

    deinit {
        WalletC.Wallet.delete(wallet: self.pointer)
    }
}
